system/hardware/tici/agnos.py

```diff
@@ -172,9 +165,68 @@ def flash_partition(target_slot_number: int, partition: dict, cloudlog):
     if out.tell() != partition['size']:
       raise Exception("Uncompressed size mismatch")
 
-    # Write hash after successfull flash
     os.sync()
-    if not full_check:
+
+
+def extract_casync_image(target_slot_number: int, partition: dict, cloudlog):
+  path = get_partition_path(target_slot_number, partition)
+  seed_path = path[:-1] + ('b' if path[-1] == 'a' else 'a')
+
+  target = casync.parse_caibx(partition['casync_caibx'])
+
+  sources = []
+
+  # First source is the current partition. Index file for current version is provided in the manifest
+  # TODO: build url based on agnos version or parition hash instead?
+  if 'casync_seed_caibx' in partition:
+    sources += [('seed', functools.partial(casync.read_chunk_local_file, f=open(seed_path, 'rb')), casync.build_chunk_dict(casync.parse_caibx(partition['casync_seed_caibx'])))]
+
+  # Second source is the target partition, this allows for resuming
+  sources += [('target', functools.partial(casync.read_chunk_local_file, f=open(path, 'rb')), casync.build_chunk_dict(target))]
+
+  # Finally we add the remote source to download any missing chunks
+  sources += [('remote', functools.partial(casync.read_chunk_remote_store, store_path=partition['casync_store']), casync.build_chunk_dict(target))]
+
+  last_p = 0
+
+  def progress(cur):
+    nonlocal last_p
+    p = int(cur / partition['size'] * 100)
+    if p != last_p:
+      last_p = p
+      print(f"Installing {partition['name']}: {p}", flush=True)
+
+  stats = casync.extract(target, sources, path, progress)
+  cloudlog.error(f'casync done {json.dumps(stats)}')
```

Comment:
```suggestion
  cloudlog.event(f'casync done', stats=stats, error=True)
```

Reply:
We pass in a normal logger in standalone mode, which doesn't have `event`